base/compiler/ssair/passes.jl

```diff
@@ -1880,6 +1880,196 @@ function adce_pass!(ir::IRCode, inlining::Union{Nothing,InliningState}=nothing)
     return complete(compact)
 end
 
+function arrayptr(P::Union{Type{Ptr{T}},Type{Ptr{Cvoid}}}, x::Array{T}, i::Int)::P where T
+    # Based on unsafe_convert for RefArray
+    if allocatedinline(T)
+        p = pointer(x, i)
+    elseif isconcretetype(T) && ismutabletype(T)
+        p = pointer_from_objref(x[i])
+    else
+        p = pointerref(Ptr{Ptr{Cvoid}}(pointer(x, i)), 1, Core.sizeof(Ptr{Cvoid}))
+    end
+    return p
+end
+
+struct CongruenceClass
+    ssa::Int
+    blockidx::Int
+end
+
+function perform_symbolic_evaluation(stmt::Expr, ssa_to_ssa)
+    # rename all SSAValues
+    # taken from renumber_ir_elements!
+    if stmt.head !== :enter && !is_meta_expr_head(stmt.head)
+
+        # key = similar(stmt.args, length(stmt.args)+1)
+        # copyto!(key, stmt.args)
+        # key[end] = stmt.head
+        # for (i, arg) in enumerate(stmt.args)
+        #     if isa(arg, SSAValue)
+        #         key[i] = SSAValue(ssa_to_ssa[arg.id])
+        #     end
+        # end
+        # svec(key...)
+
+        key = ccall(:jl_alloc_svec, SimpleVector, (Csize_t,), length(stmt.args)+1);
+        ptr = convert(Ptr{Ptr{Cvoid}}, pointer_from_objref(key)) + sizeof(Ptr{Cvoid})
+        for (i, arg) in enumerate(stmt.args)
+            if isa(arg, SSAValue)
+                unsafe_store!(ptr, ccall(:jl_box_ssavalue, Ptr{Any}, (Csize_t,), ssa_to_ssa[arg.id]), i)
+            else
+                unsafe_store!(ptr, arrayptr(Ptr{Nothing}, stmt.args, i), i)
+            end
+        end
+        unsafe_store!(ptr, pointer_from_objref(stmt.head), length(stmt.args)+1)
+        return key
+    else
+        return nothing
+    end
+end
+function perform_symbolic_evaluation(stmt::PhiNode, ssa_to_ssa, blockidx)
+    length(stmt.values) == 0 && return nothing
+
+    no_of_edges = length(stmt.edges)
+    key = Vector{Any}(undef, no_of_edges*2 + 2)
```

Comment:
I think for this to work properly, the edges need to be sorted. I don't think we currently have that invariant in the IR in general, though we probably could.